front-end-tools/src/reading_from_blockchain.ts

```diff
@@ -0,0 +1,146 @@
+    inputParameterType: string | undefined,
+    hasInputParameter: boolean,
+    deriveFromSmartContractIndex: boolean
+) {
+    if (rpcClient === undefined) {
+        throw new Error(`rpcClient undefined`);
+    }
+
+    if (contractName === undefined) {
+        throw new Error(`Set contract name`);
+    }
+
+    if (entryPoint === undefined) {
+        throw new Error(`Set entry point name`);
+    }
+
+    let param = toBuffer('', 'hex');
+
+    if (hasInputParameter) {
+        if (!deriveFromSmartContractIndex && moduleSchema === undefined) {
+            throw new Error(`Set schema`);
+        } else if (deriveFromSmartContractIndex && moduleSchema === undefined) {
+            throw new Error(`No embedded module schema found in module`);
+        }
+
+        if (inputParameterType === undefined) {
+            throw new Error(`Select input parameter type`);
+        }
+
+        let inputParameterFormated;
+
+        if (inputParameter === undefined) {
+            throw new Error(`Set input parameter`);
+        }
+
+        switch (inputParameterType) {
+            case 'number':
+                inputParameterFormated = Number(inputParameter);
+                break;
+            case 'string':
+                inputParameterFormated = inputParameter;
+                break;
+            case 'object':
+                inputParameterFormated = JSON.parse(inputParameter);
+                break;
+            case 'array':
+                inputParameterFormated = JSON.parse(inputParameter);
+                break;
+            default:
+                throw new Error(`InputParameterType does not exist`);
+        }
+
+        if (moduleSchema !== undefined) {
+            param = serializeUpdateContractParameters(
+                contractName,
+                entryPoint,
+                inputParameterFormated,
+                toBuffer(moduleSchema, 'base64')
+            );
+        }
+    }
+
+    const res = await rpcClient.invokeContract({
+        method: `${contractName}.${entryPoint}`,
+        contract: { index: contractIndex, subindex: CONTRACT_SUB_INDEX },
+        parameter: param,
+    });
+
+    if (!res || res.tag === 'failure' || !res.returnValue) {
+        throw new Error(
+            `RPC call 'invokeContract' on method '${contractName}.${entryPoint}' of contract '${contractIndex}' failed`
+        );
+    }
+
+    if (moduleSchema === undefined) {
+        // If no schema is provided return the raw bytes
+        return JSON.stringify(res.returnValue);
+    }
+
+    // If schema is provided deserialize return value
+    const returnValue = deserializeReceiveReturnValue(
+        toBuffer(res.returnValue, 'hex'),
+        toBuffer(moduleSchema, 'base64'),
+        contractName,
+        entryPoint
+    );
+
+    if (returnValue === undefined) {
+        throw new Error(
+            `Deserializing the returnValue from the '${contractName}.${entryPoint}' method of contract '${contractIndex}' failed`
+        );
```

Comment:
Don't you get any useful error message here?

Reply:
I return the original response object/error as well now.

Reply:
If the function returns no `return_value`, it is an error for me (miss use of this box). Dry running should be implemented before executing the `init/deploy/update` transaction in the other boxes instead.

Reply:
The error is now returned as well from the `deserializeReceiveReturnValue` function.
```

    let returnValue;

    try {
        // If schema is provided deserialize return value
        returnValue = deserializeReceiveReturnValue(
            toBuffer(res.returnValue, 'hex'),
            toBuffer(moduleSchema, 'base64'),
            contractName,
            entryPoint
        )
    } catch (e) {
        throw new Error(
            `Deserializing the returnValue from the '${contractName}.${entryPoint}' method of contract '${contractIndex}' failed. Original: ${e}`
        );
    }

    if (returnValue === undefined) {
        throw new Error(
            `Deserializing the returnValue from the '${contractName}.${entryPoint}' method of contract '${contractIndex}' failed.`
        );
    } else {
        return JSON.stringify(returnValue);
    }
```